lib/OpenQA/Schema/Result/ScheduledProducts.pm

```diff
@@ -565,22 +569,8 @@ sub _generate_jobs {
 
             # variable expansion
             # replace %NAME% with $settings{NAME}
-            my $expanded;
-            do {
-                $expanded = 0;
-                for my $var (keys %settings) {
-                    my $val = $settings{$var};
-                    next unless ($val && $val =~ /(%\w+%)/);
-                    my $replace_var = $1;
-                    $replace_var =~ s/^%(\w+)%$/$1/;
-                    my $replace_val = $settings{$replace_var};
-                    next unless (defined $replace_val);
-                    $replace_val = '' if $replace_var eq $var;    #stop infinite recursion
-                    $val =~ s/%${replace_var}%/$replace_val/g;
-                    $settings{$var} = $val;
-                    $expanded = 1;
-                }
-            } while ($expanded);
+            my $error = OpenQA::ExpandPlaceholder::expand_placeholders(\%settings);
+            $error_message .= $error if defined $error;
```

Comment:
That wouldn't insert a whitespace or any kind of separation between the error messages. Better add the errors to an array first and join it in the end (maybe even in the very end when rendering the error message to JSON).

Reply:
I checked the usage of `die`. It says that `If the string exception does not end in a newline, the current script line number and input line number (if any) and a newline are appended to it.`
Previous, the error_message is such as:
`Error: The key BUILD_HA contains a circular reference, its value is %BUILD% at lib/OpenQA/ExpandPlaceholder.pm line 44.
Error: The key BUILD_HA contains a circular reference, its value is %BUILD% at lib/OpenQA/ExpandPlaceholder.pm line 44.`
I think maybe it not suitable to let the web page to show the line where error happens.  So I add `\n` in the end of the line. The error message is such as:
`Error: The key BUILD_HA contains a circular reference, its value is %BUILD%.
 Error: The key BUILD_HA contains a circular reference, its value is %BUILD%.`

Do you think which is better?

Reply:
Ah ok. The one without the file and line number is better since it is not an unexpected error.